synapse/assets/storm/migrations/model-0.2.31.storm

```diff
@@ -0,0 +1,511 @@
+                }
+            }
+        }
+    }
+
+    // This loop stores all the edits and edges from every layer
+    for $rview in $rviews {
+        if (not $layers.has($rview.layers.0.iden)) { continue }
+
+        view.exec $rview.iden {
+            yield $iden
+            delnode --deledges --force
+        }
+    }
+
+    return()
+}
+
+function migrateCpeNode(oldcpe, newcpe) {
+
+    $refinfos = $getRefInfo($oldcpe.form())
+
+    // Iterate through the views and fix up all the references
+    for $view in $views {
+        if (not $layers.has($view.layers.0.iden)) { continue }
+
+        view.exec $view.iden {
+
+            $_viewiden = $lib.view.get().iden
+            $_layriden = $lib.layer.get().iden
+
+            // Fix references that point to old node to now point to new node
+            for $refinfo in $refinfos {
+                ($form, $prop, $proptype, $isarray, $isro) = $refinfo
+                if ($proptype = "ndef") {
+                    $oldvalu = $oldcpe.ndef()
+                    $newvalu = $newcpe.ndef()
+                } else {
+                    $oldvalu = $oldcpe.repr()
+                    $newvalu = $newcpe.repr()
+                }
+
+                for $n in $lib.model.migration.liftByPropValuNoNorm($form, $prop, $oldvalu) {
+                    yield $n
+
+                    if $isro {
+                        // The property is readonly so we can only delete it
+                        $lib.log.warning(`{$form}:{$prop} is readonly, cannot modify. Storing node: {$node.repr()}`)
+                        $removeNode($node, $_viewiden, $_layriden)
+                        | delnode
+                        continue
+                    }
+
+                    if $isarray {
+
+                        // We can't just [ :$prop-=$oldvalu :$prop+=$newvalu ] because the norm() function gets called
+                        // on the array type deep down in the AST. So, instead, we have to operate on the whole array.
+
+                        $list = $lib.copy(:$prop)
+                        $list.rem($oldvalu, all=$lib.true)
+
+                        if (not $list.has($newvalu)) {
+                            $list.append($newvalu)
+                        }
+
+                        $lib.model.migration.setNodePropValuNoNorm($node, $prop, $list)
+
+                    } else {
+                        [ -:$prop :$prop=$newvalu ]
+                    }
+                }
+            }
+        }
+    }
+
+    return()
+}
+
+// This is a separate function because I was getting weird bugs from scoping
+// issues. Mainly, the $view variable being overwritten because this code was
+// scoped to the whole script.
+function main() {
+
+    // Delete invalid floating :v2_2 props in all views
+    for $rview in $rviews {
+        if (not $layers.has($rview.layers.0.iden)) { continue }
+
+        view.exec $rview.iden {
+
+            for $sode in $rview.layers.0.getStorNodesByForm(it:sec:cpe) {
```

Comment:
Can we use `$_layer = $lib.layer.get()` to get the current write layer without having to rely on the heavy object?